### ISSUE TITLE
Add Octopath Traveler bingo

### ISSUE THREAD

**Issue by CAD97** [kbuzsaki/bingosync#46]
I'm not exactly certain what all has to be added other than the generator, so I'm making this an issue rather than a PR. Due to the structure - 8 sets of goals which don't want to be in the same bingo in order to make for a somewhat balanced card around the length of All Prologues - it required a custom generator, which I've implemented a version of below.

I actually copped out on trying to generate a unique layout so just hardcoded one that gets randomly mirrored and transposed and filled in to hide the pattern a bit. I'll probably come back to the problem later and figure out more of these not-quite-magic-squares or at least how to generate one in no worse than quadratic time (so I can hardcode more options).

```javascript
// What even is this magic?
(function(j, i, g, m, k, n, o) {
    function q(b) {
        var e, f, a = this,
            c = b.length,
            d = 0,
            h = a.i = a.j = a.m = 0;
        a.S = [];
        a.c = [];
        for (c || (b = [c++]); d < g;) a.S[d] = d++;
        for (d = 0; d < g; d++) e = a.S[d], h = h + e + b[d % c] & g - 1, f = a.S[h], a.S[d] = f, a.S[h] = e;
        a.g = function(b) {
            var c = a.S,
                d = a.i + 1 & g - 1,
                e = c[d],
                f = a.j + e & g - 1,
                h = c[f];
            c[d] = h;
            c[f] = e;
            for (var i = c[e + h & g - 1]; --b;) d = d + 1 & g - 1, e = c[d], f = f + e & g - 1, h = c[f], c[d] = h, c[f] = e, i = i * g + c[e + h & g - 1];
            a.i = d;
            a.j = f;
            return i
        };
        a.g(g)
    }

    function p(b, e, f, a, c) {
        f = [];
        c = typeof b;
        if (e && c == "object")
            for (a in b)
                if (a.indexOf("S") < 5) try {
                    f.push(p(b[a], e - 1))
                } catch (d) {}
                return f.length ? f : b + (c != "string" ? "\0" : "")
    }

    function l(b, e, f, a) {
        b += "";
        for (a = f = 0; a < b.length; a++) {
            var c = e,
                d = a & g - 1,
                h = (f ^= e[a & g - 1] * 19) + b.charCodeAt(a);
            c[d] = h & g - 1
        }
        b = "";
        for (a in e) b += String.fromCharCode(e[a]);
        return b
    }
    i.seedrandom = function(b, e) {
        var f = [],
            a;
        b = l(p(e ? [b, j] : arguments.length ? b : [(new Date).getTime(), j, window], 3), f);
        a = new q(f);
        l(a.S, j);
        i.random = function() {
            for (var c = a.g(m), d = o, b = 0; c < k;) c = (c + b) * g, d *= g, b = a.g(1);
            for (; c >= n;) c /= 2, d /= 2, b >>>= 1;
            return (c + b) / d
        };
        return b
    };
    o = i.pow(g, m);
    k = i.pow(2, k);
    n = k * 2;
    l(i.random(), j)
})([], Math, 256, 6, 52);

bingoGenerator = function(bingoList, opts) {
  const LANG = opts.lang || 'name';
  const MODE = opts.mode || 'normal';
  const cardType = 'Normal';
  const SEED = opts.seed || Math.ceil(999999 * Math.random()).toString();
  const size = 5;
  
  // octopath generator by @CAD97
  Math.seedrandom(SEED);
  
  function shuffle(list) {
    for (let i = list.length - 1; i > 0; i--) {
      const j = Math.floor(Math.random() * (i + 1));
      [list[i], list[j]] = [list[j], list[i]];
    }
    return list;
  }
  function transpose(matrix) {
    for (let i = 0; i < matrix.length; i++) {
      for (let j = 0; j < i; j++) {
        [matrix[i][j], matrix[j][i]] = [matrix[j][i], matrix[i][j]]
      }
    }
    return matrix;
  }
  function horizontalMirror(matrix) {
    for (let i = 0; i < matrix.length; i++) {
      matrix[i] = matrix[i].reverse();
    }
    return matrix;
  }
  function verticalMirror(matrix) {
    return matrix.reverse();
  }
  function coinFlip() {
    return Math.random() < 0.5;
  }
  function randomPop(arr) {
    const i = Math.floor(Math.random() * arr.length);
    const ret = arr[i];
    arr.splice(i, 1);
    return ret;
  }
  
  let boardLayout = [
    [0, 2, 4, 6, 7],
    [1, 3, 2, 5, 4],
    [3, 4, 1, 2, 6],
    [5, 6, 3, 7, 0],
    [2, 1, 7, 0, 5]
  ];
  if (coinFlip()) boardLayout = transpose(boardLayout);
  if (coinFlip()) boardLayout = horizontalMirror(boardLayout);
  if (coinFlip()) boardLayout = verticalMirror(boardLayout);
  bingoList = shuffle(bingoList);
  
  return [].concat(...boardLayout.map(row =>
    row.map(idx => randomPop(bingoList[idx]))));
}

var bingoList = [];
bingoList[0] = [
	{ "name": "Ophilia" },
	{ "name": "Ophilia - Sheltering Veil" },
	{ "name": "Ophilia - Luminescence" },
	{ "name": "Ophilia - Heal More" },
	{ "name": "Ophilia - Reflective Veil" },
	{ "name": "Ophilia - Revive" }
];
bingoList[1] = [
	{ "name": "Cyrus" },
	{ "name": "Cyrus - Lightning Bolt" },
	{ "name": "Cyrus - Analyze" },
	{ "name": "Cyrus - Fire Storm" },
	{ "name": "Cyrus - Blizzard" },
	{ "name": "Cyrus - Lightning Blast" }
];
bingoList[2] = [
	{ "name": "Tressa" },
	{ "name": "Tressa - Rest" },
	{ "name": "Tressa - Trade Tempest" },
	{ "name": "Tressa - Donate BP" },
	{ "name": "Tressa - Sidestep" },
	{ "name": "Tressa - Hired Help" }
];
bingoList[3] = [
	{ "name": "Olberic" },
	{ "name": "Olberic - Spearhead" },
	{ "name": "Olberic - Incite" },
	{ "name": "Olberic - Cross Strike" },
	{ "name": "Olberic - Stout Wall" },
	{ "name": "Olberic - Thousand Spears" }
];
bingoList[4] = [
	{ "name": "Primrose" },
	{ "name": "Primrose - Peacock Strut" },
	{ "name": "Primrose - Mole Dance" },
	{ "name": "Primrose - Night Ode" },
	{ "name": "Primrose - Panther Dance" },
	{ "name": "Primrose - Bewildering Grace" }
];
bingoList[5] = [
	{ "name": "Alfyn" },
	{ "name": "Alfyn - Rehabilitate" },
	{ "name": "Alfyn - Amputation" },
	{ "name": "Alfyn - Empoison" },
	{ "name": "Alfyn - Vivify" },
	{ "name": "Alfyn - Last Stand" }
];
bingoList[6] = [
	{ "name": "Therion" },
	{ "name": "Therion - HP Thief" },
	{ "name": "Therion - Shackle Foe" },
	{ "name": "Therion - Armor Corrosive" },
	{ "name": "Therion - Steal SP" },
	{ "name": "Therion - Share SP" }
];
bingoList[7] = [
	{ "name": "H'aanit" },
	{ "name": "H'aanit - Thunderbird" },
	{ "name": "H'aanit - Leghold Trap" },
	{ "name": "H'aanit - Mercy Strike" },
	{ "name": "H'aanit - Arrowstorm" },
	{ "name": "H'aanit - Take Aim" }
];
```

I used a few more recent javascriptisms but I can dumb it down if you want as well.

**Comment by CAD97** [comment]
Example outputs:

```json


[
  {
    "name": "Primrose - Bewildering Grace"
  },
  {
    "name": "Ophilia - Luminescence"
  },
  {
    "name": "Tressa - Collect"
  },
  {
    "name": "Alfyn - Last Stand"
  },
  {
    "name": "Cyrus - Analyze"
  },
  {
    "name": "Tressa - Donate BP"
  },
  {
    "name": "Cyrus - Blizzard"
  },
  {
    "name": "H'aanit - Mercy Strike"
  },
  {
    "name": "Primrose - Peacock Strut"
  },
  {
    "name": "Alfyn - First Aid"
  },
  {
    "name": "Ophilia - Revive"
  },
  {
    "name": "H'aanit - Rain of Arrows"
  },
  {
    "name": "Therion - Steal SP"
  },
  {
    "name": "Olberic - Thousand Spears"
  },
  {
    "name": "Primrose - Panther Dance"
  },
  {
    "name": "H'aanit - True Strike"
  },
  {
    "name": "Olberic - Level Slash"
  },
  {
    "name": "Ophilia - Heal Wounds"
  },
  {
    "name": "Tressa - Sidestep"
  },
  {
    "name": "Therion - Share SP"
  },
  {
    "name": "Alfyn - Icicle"
  },
  {
    "name": "Therion"
  },
  {
    "name": "Olberic - Spearhead"
  },
  {
    "name": "Cyrus"
  },
  {
    "name": "H'aanit - Arrowstorm"
  }
]
```

```json


[
  {
    "name": "Ophilia - Sheltering Veil"
  },
  {
    "name": "H'aanit"
  },
  {
    "name": "Tressa - Rest"
  },
  {
    "name": "Primrose"
  },
  {
    "name": "Cyrus - Analyze"
  },
  {
    "name": "H'aanit - True Strike"
  },
  {
    "name": "Tressa - Donate BP"
  },
  {
    "name": "Alfyn"
  },
  {
    "name": "Therion - Steal"
  },
  {
    "name": "Ophilia - Heal Wounds"
  },
  {
    "name": "Therion - Wildfire"
  },
  {
    "name": "Cyrus - Lightning Blast"
  },
  {
    "name": "Primrose - Panther Dance"
  },
  {
    "name": "Olberic"
  },
  {
    "name": "Alfyn - Empoison"
  },
  {
    "name": "Olberic - Abide"
  },
  {
    "name": "Ophilia - Heal More"
  },
  {
    "name": "Cyrus - Lightning Bolt"
  },
  {
    "name": "Alfyn - Rehabilitate"
  },
  {
    "name": "Primrose - Lion Dance"
  },
  {
    "name": "Tressa - Sidestep"
  },
  {
    "name": "Therion - Steal SP"
  },
  {
    "name": "Olberic - Incite"
  },
  {
    "name": "Cyrus - Fire Storm"
  },
  {
    "name": "H'aanit - Arrowstorm"
  }
]
```

**Comment by CAD97** [comment]
I'm not a primary JS dev so I don't really know how to use jsfiddle, but [this should be a link to a working demonstration of the provided code](https://jsfiddle.net/y180zwh2/93/).

**Comment by CAD97** [comment]
I just updated the goals to remove accidentally duplicated goals (which didn't look duplicated, but were -- "acquire skill" for a character where they start with that skill, and "acquire character" is a goal already).

**Comment by kbuzsaki** [comment]
Thanks for filling the issue-- I'll be busy tonight and tomorrow, but I should be able to take a look some time on the 4th or 5th.

**Comment by kbuzsaki** [comment]
Added your generator in 5e7c73258a86805da6e7848afe3daaa3f1b4eea5 and dumbed down the ES6-isms in 2f02b645f13768aa8b261e7f9f42606f6d880eb0. I verified that they produce the same cards.

Up on the site now, let me know if you'd like to update it :)

**Comment by kbuzsaki** [comment]
Oh, also, I abbreviated the "short name" to "Octopath Trav.". The longer name shows up in the game selection list, the short name shows up everywhere else-- the active rooms panel, the history page, the room settings panel within a room.

I made this abbreviation to prevent it from line breaking as often. Let me know if you'd prefer to have a different abbreviation or no abbreviation at all.

**Comment by CAD97** [comment]
Thanks for putting it up! The bingo sets will probably evolve when the game's actually out, I'll PR for that.

The normal short name is just "Octopath".

**Comment by kbuzsaki** [comment]
Updated the short name to be "Octopath"
